File: OpenBurnBarCore/Sources/OpenBurnBarCore/Hermes/HermesAtomURL.swift
import Foundation

// MARK: - Hermes Atom URL Codec
//
// Canonical `burnbar://` URL encoding for `HermesAtom`. The same vocabulary
// is documented to Hermes via `HermesSystemPromptBuilder` so the model emits
// matching markdown links the client can decode.
//
// URL forms:
//   burnbar://burn?window=today&amount=2.34
//   burnbar://session?id=abc-123
//   burnbar://provider?token=anthropic
//   burnbar://model?id=claude-sonnet-4.7
//   burnbar://window?value=7d
//   burnbar://tool?name=ReadFile
//   burnbar://project?id=BurnBar
//   burnbar://tokens?value=12400&scope=today
//   burnbar://quota?provider=anthropic&percent=78
//   burnbar://runtime?profile=hermes

/// Canonical scheme used for in-app navigation links emitted by Hermes.
public let HermesAtomURLScheme = "burnbar"

public enum HermesAtomURL {
    private static let maxURLLength = 2_048
    private static let maxShortPayloadLength = 256
    private static let maxProjectPayloadLength = 512
    private static let maxTokenValue = 1_000_000_000
    private static let maxCostAmount = 1_000_000_000.0

    private static let providerCharacterSet = CharacterSet(charactersIn: "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789._-")
    private static let identifierCharacterSet = CharacterSet(charactersIn: "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789._:+/-")

    /// Encode a `HermesAtom` to its canonical `burnbar://` URL.
    public static func encode(_ atom: HermesAtom) -> URL {
        var components = URLComponents()
        components.scheme = HermesAtomURLScheme
        switch atom {
        case .cost(let amount, let window):
            components.host = "burn"
            components.queryItems = [
                URLQueryItem(name: "window", value: window.rawValue),
                URLQueryItem(name: "amount", value: String(amount))
            ]
        case .session(let id):
            components.host = "session"
            components.queryItems = [URLQueryItem(name: "id", value: id)]
        case .provider(let token):
            components.host = "provider"
            components.queryItems = [URLQueryItem(name: "token", value: token)]
        case .model(let id):
            components.host = "model"
            components.queryItems = [URLQueryItem(name: "id", value: id)]
        case .window(let value):
            components.host = "window"
            components.queryItems = [URLQueryItem(name: "value", value: value.rawValue)]
        case .tool(let name):
            components.host = "tool"
            components.queryItems = [URLQueryItem(name: "name", value: name)]
        case .project(let id):
            components.host = "project"
            components.queryItems = [URLQueryItem(name: "id", value: id)]
        case .tokens(let value, let scope):
            components.host = "tokens"
            components.queryItems = [
                URLQueryItem(name: "value", value: String(value)),
                URLQueryItem(name: "scope", value: scope.rawValue)
            ]
        case .quota(let provider, let percent):
            components.host = "quota"
            components.queryItems = [
                URLQueryItem(name: "provider", value: provider),
                URLQueryItem(name: "percent", value: String(percent))
            ]
        case .runtime(let profile):
            components.host = "runtime"
            components.queryItems = [URLQueryItem(name: "profile", value: profile)]
        }
        // URLComponents builds the right form even when host has no path.
        return components.url ?? URL(string: "\(HermesAtomURLScheme)://unknown")!
    }

    /// Decode a `URL` back to a `HermesAtom`. Returns `nil` for any URL
    /// that's not a recognized burnbar:// atom — callers should fall back
    /// to rendering the link as plain text.
    public static func decode(_ url: URL) -> HermesAtom? {
        guard url.scheme?.lowercased() == HermesAtomURLScheme else { return nil }
        guard url.absoluteString.count <= maxURLLength else { return nil }
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else { return nil }
        guard components.user == nil, components.password == nil, components.fragment == nil else { return nil }
        guard components.path.isEmpty || components.path == "/" else { return nil }
        let host = (components.host ?? "").lowercased()
        guard let params = decodeQueryItems(components.queryItems ?? []) else { return nil }
        return decode(host: host, params: params)
    }

    /// Decode a `String` URL form to a `HermesAtom`. Used by the Markdown
    /// link parser to avoid double-allocating `URL` objects.
    public static func decode(_ urlString: String) -> HermesAtom? {
        guard urlString.count <= maxURLLength else { return nil }
        guard let url = URL(string: urlString) else { return nil }
        return decode(url)
    }

    private static func decodeQueryItems(_ items: [URLQueryItem]) -> [String: String]? {
        var params: [String: String] = [:]
        for item in items {
            let name = item.name.lowercased()
            guard !name.isEmpty, name.count <= maxShortPayloadLength else { return nil }
            guard params[name] == nil else { return nil }
            guard let value = item.value, value.count <= maxProjectPayloadLength else { return nil }
            guard !containsControlCharacters(name), !containsControlCharacters(value) else { return nil }
            params[name] = value
        }
        return params
    }

    private static func decode(host: String, params: [String: String]) -> HermesAtom? {
        switch host {
        case "burn":
            let window = params["window"].flatMap(HermesAtomWindow.init(rawValue:)) ?? .today
            let amount = params["amount"].flatMap(Double.init) ?? 0
            guard amount.isFinite, amount >= 0, amount <= maxCostAmount else { return nil }
            return .cost(amount: amount, window: window)
        case "session":
            guard let id = params["id"], isOpaquePayload(id, maxLength: maxShortPayloadLength) else { return nil }
            return .session(id: id)
        case "provider":
            guard let token = params["token"], isProviderToken(token) else { return nil }
            return .provider(token: token)
        case "model":
            guard let id = params["id"], isIdentifierToken(id) else { return nil }
            return .model(id: id)
        case "window":
            guard let value = params["value"].flatMap(HermesAtomWindow.init(rawValue:)) else { return nil }
            return .window(value)
        case "tool":
            guard let name = params["name"], isIdentifierToken(name) else { return nil }
            return .tool(name: name)
        case "project":
            guard let id = params["id"], isOpaquePayload(id, maxLength: maxProjectPayloadLength) else { return nil }
            return .project(id: id)
        case "tokens":
            guard let raw = params["value"], let value = Int(raw) else { return nil }
            guard (0...maxTokenValue).contains(value) else { return nil }
            let scope = params["scope"].flatMap(HermesAtomTokenScope.init(rawValue:)) ?? .unspecified
            return .tokens(value: value, scope: scope)
        case "quota":
            guard let provider = params["provider"], !provider.isEmpty,
                  let percentRaw = params["percent"], let percent = Int(percentRaw) else { return nil }
            guard isProviderToken(provider), (0...100).contains(percent) else { return nil }
            return .quota(provider: provider, percent: percent)
        case "runtime":
            guard let profile = params["profile"], isIdentifierToken(profile) else { return nil }
            return .runtime(profile: profile)
        default:
            return nil
        }
    }

    private static func isProviderToken(_ value: String) -> Bool {
        isASCIISet(value, maxLength: maxShortPayloadLength, allowed: providerCharacterSet) && !value.contains("..")
    }

    private static func isIdentifierToken(_ value: String) -> Bool {
        isASCIISet(value, maxLength: maxShortPayloadLength, allowed: identifierCharacterSet) && !value.contains("..")
    }

    private static func isASCIISet(_ value: String, maxLength: Int, allowed: CharacterSet) -> Bool {
        guard !value.isEmpty, value.count <= maxLength, !containsControlCharacters(value) else { return false }
        return value.unicodeScalars.allSatisfy { $0.isASCII && allowed.contains($0) }
    }

    private static func isOpaquePayload(_ value: String, maxLength: Int) -> Bool {
        guard !value.isEmpty, value.count <= maxLength, !containsControlCharacters(value) else { return false }
        guard !value.contains("://"), !value.contains("\u{0000}") else { return false }
        let parts = value.split(separator: "/", omittingEmptySubsequences: false)
        return !parts.contains { $0 == ".." }
    }

    private static func containsControlCharacters(_ value: String) -> Bool {
        value.unicodeScalars.contains { CharacterSet.controlCharacters.contains($0) }
    }
}

File: windows/app/OpenBurnBar.App.Presentation/Chat/HermesAtomUrl.cs
using System;
using System.Collections.Generic;
using System.Globalization;

namespace OpenBurnBar.App.Presentation.Chat;

// MARK: - Hermes Atom URL Codec
//
// C# peer of `OpenBurnBarCore/Sources/OpenBurnBarCore/Hermes/HermesAtomURL.swift`.
//
// Canonical `burnbar://` URL encoding for `HermesAtom`. The same vocabulary is
// documented to Hermes via the system prompt builder so the model emits matching
// markdown links the client can decode:
//
//   burnbar://burn?window=today&amount=2.34
//   burnbar://session?id=abc-123
//   burnbar://provider?token=anthropic
//   burnbar://model?id=claude-sonnet-4.7
//   burnbar://window?value=7d
//   burnbar://tool?name=ReadFile
//   burnbar://project?id=BurnBar
//   burnbar://tokens?value=12400&scope=today
//   burnbar://quota?provider=anthropic&percent=78
//   burnbar://runtime?profile=hermes

public static class HermesAtomUrl
{
    /// Canonical scheme used for in-app navigation links emitted by Hermes.
    public const string Scheme = "burnbar";

    /// Encode a <see cref="HermesAtom"/> to its canonical `burnbar://` URL string.
    /// Mirrors the Swift `URLComponents` builder (host = kind, query = payload).
    public static string Encode(HermesAtom atom)
    {
        switch (atom)
        {
            case HermesAtom.Cost c:
                return Build("burn", ("window", c.Window.ToRawValue()),
                    ("amount", c.Amount.ToString("R", CultureInfo.InvariantCulture)));
            case HermesAtom.Session s:
                return Build("session", ("id", s.Id));
            case HermesAtom.ProviderRef p:
                return Build("provider", ("token", p.Token));
            case HermesAtom.Model m:
                return Build("model", ("id", m.Id));
            case HermesAtom.WindowRef w:
                return Build("window", ("value", w.Value.ToRawValue()));
            case HermesAtom.Tool t:
                return Build("tool", ("name", t.Name));
            case HermesAtom.Project pr:
                return Build("project", ("id", pr.Id));
            case HermesAtom.Tokens tk:
                return Build("tokens", ("value", tk.Value.ToString(CultureInfo.InvariantCulture)),
                    ("scope", tk.Scope.ToRawValue()));
            case HermesAtom.Quota q:
                return Build("quota", ("provider", q.Provider),
                    ("percent", q.Percent.ToString(CultureInfo.InvariantCulture)));
            case HermesAtom.Runtime r:
                return Build("runtime", ("profile", r.Profile));
            default:
                return $"{Scheme}://unknown";
        }
    }

    /// Decode a URL string back to a <see cref="HermesAtom"/>. Returns
    /// <c>null</c> for any string that is not a recognized `burnbar://` atom —
    /// callers fall back to rendering the link as plain text. Byte-for-byte port
    /// of Swift `HermesAtomURL.decode`.
    public static HermesAtom? Decode(string? urlString)
    {
        if (string.IsNullOrEmpty(urlString))
        {
            return null;
        }
        if (!Uri.TryCreate(urlString, UriKind.Absolute, out var uri))
        {
            return null;
        }
        if (!string.Equals(uri.Scheme, Scheme, StringComparison.OrdinalIgnoreCase))
        {
            return null;
        }

        // Swift reads the URL "host" as the atom kind. For `scheme://host?query`
        // the .NET Uri exposes it as Host; guard against forms that route the
        // kind into the path (e.g. `burnbar:///burn`).
        var host = uri.Host;
        if (string.IsNullOrEmpty(host))
        {
            return null;
        }
        host = host.ToLowerInvariant();

        var parms = UniqueQueryParams(uri.Query);
        if (parms is null)
        {
            return null;
        }
        return Decode(host, parms);
    }

    private static HermesAtom? Decode(string host, IReadOnlyDictionary<string, string> parms)
    {
        switch (host)
        {
            case "burn":
            {
                var window = (parms.TryGetValue("window", out var wRaw)
                    ? HermesAtomWindowRaw.FromRawValue(wRaw) : null) ?? HermesAtomWindow.Today;
                var amount = parms.TryGetValue("amount", out var aRaw)
                    && double.TryParse(aRaw, NumberStyles.Float, CultureInfo.InvariantCulture, out var parsed)
                    ? parsed : 0;
                return new HermesAtom.Cost(amount, window);
            }
            case "session":
                return NonEmpty(parms, "id") is { } id ? new HermesAtom.Session(id) : null;
            case "provider":
                return NonEmpty(parms, "token") is { } token ? new HermesAtom.ProviderRef(token) : null;
            case "model":
                return NonEmpty(parms, "id") is { } mid ? new HermesAtom.Model(mid) : null;
            case "window":
            {
                var value = parms.TryGetValue("value", out var vRaw)
                    ? HermesAtomWindowRaw.FromRawValue(vRaw) : null;
                return value is { } v ? new HermesAtom.WindowRef(v) : null;
            }
            case "tool":
                return NonEmpty(parms, "name") is { } name ? new HermesAtom.Tool(name) : null;
            case "project":
                return NonEmpty(parms, "id") is { } pid ? new HermesAtom.Project(pid) : null;
            case "tokens":
            {
                if (!parms.TryGetValue("value", out var raw)
                    || !int.TryParse(raw, NumberStyles.Integer, CultureInfo.InvariantCulture, out var value))
                {
                    return null;
                }
                var scope = (parms.TryGetValue("scope", out var sRaw)
                    ? HermesAtomTokenScopeRaw.FromRawValue(sRaw) : null) ?? HermesAtomTokenScope.Unspecified;
                return new HermesAtom.Tokens(value, scope);
            }
            case "quota":
            {
                var provider = NonEmpty(parms, "provider");
                if (provider is null
                    || !parms.TryGetValue("percent", out var pctRaw)
                    || !int.TryParse(pctRaw, NumberStyles.Integer, CultureInfo.InvariantCulture, out var percent))
                {
                    return null;
                }
                return new HermesAtom.Quota(provider, percent);
            }
            case "runtime":
                return NonEmpty(parms, "profile") is { } profile ? new HermesAtom.Runtime(profile) : null;
            default:
                return null;
        }
    }

    // Reject duplicate query keys (matches the Swift `uniqueQueryParams` guard:
    // a repeated key returns nil for the whole URL). Keys are lower-cased.
    private static Dictionary<string, string>? UniqueQueryParams(string query)
    {
        var result = new Dictionary<string, string>(StringComparer.Ordinal);
        var trimmed = query.StartsWith("?", StringComparison.Ordinal) ? query.Substring(1) : query;
        if (trimmed.Length == 0)
        {
            return result;
        }
        foreach (var pair in trimmed.Split('&'))
        {
            if (pair.Length == 0)
            {
                continue;
            }
            var eq = pair.IndexOf('=');
            if (eq < 0)
            {
                continue; // no value — Swift skips items whose value is nil.
            }
            var name = Uri.UnescapeDataString(pair.Substring(0, eq)).ToLowerInvariant();
            var value = Uri.UnescapeDataString(pair.Substring(eq + 1));
            if (result.ContainsKey(name))
            {
                return null;
            }
            result[name] = value;
        }
        return result;
    }

    private static string? NonEmpty(IReadOnlyDictionary<string, string> parms, string key) =>
        parms.TryGetValue(key, out var value) && value.Length > 0 ? value : null;

    private static string Build(string host, params (string Name, string Value)[] query)
    {
        var builder = new System.Text.StringBuilder();
        builder.Append(Scheme).Append("://").Append(host);
        for (var i = 0; i < query.Length; i++)
        {
            builder.Append(i == 0 ? '?' : '&');
            builder.Append(Uri.EscapeDataString(query[i].Name));
            builder.Append('=');
            builder.Append(Uri.EscapeDataString(query[i].Value));
        }
        return builder.ToString();
    }
}
